src/embed/connection.rs

```diff
@@ -461,13 +464,24 @@ async fn do_get(read: db::Read<'_>, req: GetRequest) -> Result<GetResponse, Stri
     })
 }
 
-async fn do_scan(read: db::Read<'_>, req: ScanRequest) -> Result<ScanResponse, String> {
-    use std::convert::TryFrom;
-    let mut res = Vec::<ScanItem>::new();
+async fn do_scan(
+    read: db::Read<'_>,
+    req: ScanRequest,
+    req_raw: JsValue,
+) -> Result<ScanResponse, ScanError> {
+    use ScanError::*;
+    let receiver: Function = Reflect::get(&req_raw, &JsValue::from_str("receiver"))
+        .map_err(|_| MissingReceiver)?
+        .dyn_into()
+        .map_err(|_| InvalidReceiver)?;
+
     for pe in read.scan((&req.opts).into()) {
-        res.push(ScanItem::try_from(pe).map_err(to_debug)?);
+        let key = JsValue::from_str(std::str::from_utf8(pe.key).unwrap());
+        let val = unsafe { Uint8Array::view(pe.val) };
+        // TODO: receiver can return to us whether to keep going!
+        receiver.call2(&JsValue::null(), &key, &val).unwrap();
```

Comment:
Ignoramus here, is it always the case that the receiver is a free function or has some other way of not needed or getting its this/self? Is there any way to misuse it?

Reply:
It's basically part of the interface that the receiver is a free function. Another of way of saying this is that the interface we provide is you send us a function we call with a null `this`. If you want some other `this` you must arrange for that. This is easy and standard in JS (via `bind()`).